digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/AbstractJdbcOperator.java

```diff
@@ -0,0 +1,224 @@
+            }
+            queryId = state.get(QUERY_ID, UUID.class);
+        }
+
+        try (JdbcConnection connection = connect(connectionConfig)) {
+            Exception statementError = connection.validateStatement(query);
+            if (statementError != null) {
+                throw new ConfigException("Given query is invalid", statementError);
+            }
+
+            if (readOnlyMode) {
+                if (downloadFile.isPresent()) {
+                    connection.executeReadOnlyQuery(query, (results) -> downloadResultsToFile(results, downloadFile.get()));
+                }
+                else {
+                    connection.executeReadOnlyQuery(query, (results) -> skipResults(results));
+                }
+                return TaskResult.defaultBuilder(request).build();
+            }
+            else {
+                String statement;
+                boolean statementMayReturnResults;
+                if (insertInto.isPresent() || createTable.isPresent()) {
+                    if (insertInto.isPresent()) {
+                        statement = connection.buildInsertStatement(query, insertInto.get());
+                    }
+                    else {
+                        statement = connection.buildCreateTableStatement(query, createTable.get());
+                    }
+
+                    Exception modifiedStatementError = connection.validateStatement(statement);
+                    if (modifiedStatementError != null) {
+                        throw new ConfigException("Given query is valid but failed to build INSERT INTO statement (this may happen if given query includes multiple statements or semicolon \";\"?)", modifiedStatementError);
+                    }
+                    statementMayReturnResults = false;
+                    logger.debug("Running a modified statement: {}", statement);
+                }
+                else {
+                    statement = query;
+                    statementMayReturnResults = true;
+                }
+
+                TransactionHelper txHelper;
+                if (strictTransaction) {
+                    txHelper = connection.getStrictTransactionHelper(statusTableName);
+                }
+                else {
+                    txHelper = new NoTransactionHelper();
+                }
+
+                txHelper.prepare();
+
+                boolean executed = txHelper.lockedTransaction(queryId, () -> {
+                    if (statementMayReturnResults) {
+                        connection.executeScript(statement);
+                    }
+                    else {
+                        connection.executeUpdate(statement);
+                    }
+                });
+
+                if (!executed) {
+                    logger.debug("Query is already completed according to status table. Skipping statement execution.");
+                }
+
+                try {
+                    txHelper.cleanup();
```

Comment:
I see. With cleaning up old garbage status records, we don't need to have the status of a query in Digdag.

But I have a concern in the following case:
1. Digdag workflow started a query`INSERT INTO sales_summary SELECT * FROM today_sales`
2. `queryId=1234` was created and stored in Digdag state
3. A status record `query_id=1234, created_at=2016-01-01 00:00:00` was inserted to status table in PostgreSQL
4. The query that includes INSERT INTO finished successfully
5. The status record was updated to `query_id=1234, created_at=2016-01-01 00:00:00, completed_at=2016-01-01 00:11:11`
6. But the Digdag agent process crashed before returning from `runTask()`
7. Also the retry of this task was somehow delayed more than 24 hours and another workflow or task removed the status record of this query from PostgreSQL
8. This query was retried after that and it executed INSERT INTO again since the status record was already removed

It's unlikely this case happens and maybe we can ignore it for now. What do you think?